onnxruntime/test/framework/tensor_test.cc

```diff
@@ -220,20 +220,33 @@ TEST(TensorTest, Strided) {
   TensorShapeVector strides{12, 4, 1};
   ASSERT_EQ(t.Shape(), shape);
   ASSERT_THAT(t.Strides(), testing::ContainerEq(gsl::make_span(strides.cbegin(), strides.cend())));
+  ASSERT_EQ(t.SizeInBytes(), sizeof(float) * 24);
   TensorShape new_shape({4, 2, 3});
   TensorShapeVector new_strides{1, 12, 4};
   t.SetShapeAndStrides(new_shape, new_strides);
   EXPECT_FALSE(t.IsContiguous());
   ASSERT_EQ(t.Shape(), new_shape);
   ASSERT_THAT(t.Strides(), testing::ContainerEq(gsl::make_span(new_strides.cbegin(), new_strides.cend())));
+  ASSERT_EQ(t.SizeInBytes(), sizeof(float) * 24);
   Tensor t2(DataTypeImpl::GetType<float>(), new_shape, data, alloc->Info(), 0L, gsl::make_span(new_strides));
   EXPECT_FALSE(t2.IsContiguous());
-  ASSERT_EQ(t.Shape(), new_shape);
+  ASSERT_EQ(t2.Shape(), new_shape);
   ASSERT_THAT(t2.Strides(), testing::ContainerEq(gsl::make_span(new_strides.cbegin(), new_strides.cend())));
+  ASSERT_EQ(t2.SizeInBytes(), sizeof(float) * 24);
   t2.SetShapeAndStrides(shape, strides);
   EXPECT_TRUE(t2.IsContiguous());
-  ASSERT_EQ(t.Shape(), new_shape);
+  ASSERT_EQ(t2.Shape(), shape);
   ASSERT_THAT(t2.Strides(), testing::ContainerEq(gsl::make_span(strides.cbegin(), strides.cend())));
+  ASSERT_EQ(t2.SizeInBytes(), sizeof(float) * 24);
+  alloc->Free(data);
+  data = alloc->Alloc(sizeof(int64_t));
+  TensorShapeVector single_element_strides{0, 0, 0};
+  Tensor t3(DataTypeImpl::GetType<int64_t>(), shape, data, alloc->Info(), 0L, gsl::make_span(single_element_strides));
```

Comment:
I tried building eager mode, and added above code to the test. But the test passed both before and after my changes. From my callstack, the "t.to(device)" doesn't contain any frame of SizeInBytes from a strided tensor. Did I miss anything to reproduce this issue? BTW, the eager mode build doesn't work if I use PyTorch 1.11 with some compile errors, so I have to downgrade my torch version, is this expected?

Reply:
Thanks for trying the test. 

Here is the PyTorch commit hash my team uses; it corresponds to a 1.11 release and builds fine. Feel free to reach out on Teams with a specific error; 1.10 should be fine for validation.

```
# commit 4f32bdd80217a51d9bb8d6c4d55b97159bc1f02d
# Author: Ke Wen <kw2501@fb.com>
# Date:   Wed Mar 16 21:46:27 2022 -0700
```

>  But the test passed both before and after my changes.

I looked into the code some more. The reason the test passed for the CPU Execution Provider in either case is that the strided copy implementation does not use `SizeInBytes()` at all: 

https://github.com/microsoft/onnxruntime/blob/efb0928e2bba9d6e942abaad63a888a75dfc976f/onnxruntime/core/framework/copy.h#L231

For zero strided tensors, only the zeroth element is accessed 

https://github.com/microsoft/onnxruntime/blob/efb0928e2bba9d6e942abaad63a888a75dfc976f/onnxruntime/core/framework/copy.h#L27

I hit a segmentation fault because I am using an external execution provider that does not use the strides information when accessing the tensor data. I will need to make a change in the external execution provider. 

>It will be good to test at the PyTorch level

I would still add the test because it provides some coverage that empty strided PyTorch tensors are copied correctly.